zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js

```diff
@@ -410,6 +415,10 @@ angular.module('zeppelinWebApp')
 
   $scope.aceChanged = function() {
     $scope.dirtyText = $scope.editor.getSession().getValue();
+
+    $timeout(function() {
+      $scope.setParagraphMode($scope.editor.getSession(), $scope.editor.getCursorPosition());
+    });
```

Comment:
@martin-g,  i have no angular experience. As backend dev i don't get exposed often to UI technologies like this and furthermore my favorite amongst them is Google Polymer. But this is another story :) 
Here i have used $timeout (with default delay of 0) as a mean to async run the evaluation. This approach has been advised by the ACE-Editor team. And indeed without the $timeout various random errors pop-up during the mode setting. 

Anyway this is a critical bit and i am not confident i've resolved it entirely. So by all means try to scrutinize it and help to do it right! ;)